File: tmp_patch_media.js
const fs2 = require('fs');
const orig = fs2.readFileSync('/var/www/integrasi-wa.jodyaryono.id/index.js', 'utf8');

// 1) Add MEDIA_BASE_URL constant after the existing WEBHOOK vars
const insertAfter = 'const WEBHOOK_ENABLED';
const idx1 = orig.indexOf(insertAfter);
const eol1 = orig.indexOf('\n', idx1);
const constBlock = "\nconst MEDIA_BASE_URL = 'https://integrasi-wa.jodyaryono.id/uploads/';\nconst UPLOADS_DIR = path.resolve('./uploads');\nif (!fs.existsSync(UPLOADS_DIR)) fs.mkdirSync(UPLOADS_DIR, { recursive: true });\n";

let patched = orig.slice(0, eol1 + 1) + constBlock + orig.slice(eol1 + 1);

// 2) Add media download logic before '    // Per-session webhook'
const mediaBlock = '\n'
    + '    // ── Download & save media ───────────────────────────────────────\n'
    + '    let mediaUrl = null;\n'
    + '    if (msg.hasMedia) {\n'
    + '        try {\n'
    + '            const media = await msg.downloadMedia();\n'
    + '            if (media && media.data) {\n'
    + "                const extMap = { 'image/jpeg': 'jpg', 'image/png': 'png', 'image/webp': 'webp', 'video/mp4': 'mp4', 'video/3gpp': '3gp', 'audio/ogg': 'ogg', 'audio/mpeg': 'mp3', 'application/pdf': 'pdf' };\n"
    + "                const ext = extMap[media.mimetype] || (media.mimetype ? media.mimetype.split('/')[1] : 'bin');\n"
    + "                const filename = 'wa_' + Date.now() + '_' + randomBytes(4).toString('hex') + '.' + ext;\n"
    + '                const filePath = path.join(UPLOADS_DIR, filename);\n'
    + "                fs.writeFileSync(filePath, Buffer.from(media.data, 'base64'));\n"
    + '                mediaUrl = MEDIA_BASE_URL + filename;\n'
    + "                console.log('[Media][' + phoneId + '] saved ' + filename + ' (' + media.mimetype + ')');\n"
    + '            }\n'
    + '        } catch (e) {\n'
    + "            console.error('[Media][' + phoneId + '] download failed:', e.message);\n"
    + '        }\n'
    + '    }\n';

const marker = '    // Per-session webhook';
patched = patched.replace(marker, mediaBlock + '\n' + marker);

// 3) Add media_url to the webhook payload - before _key
patched = patched.replace(
    '            _key: { remoteJid:',
    '            ...(mediaUrl ? { media_url: mediaUrl } : {}),\n            _key: { remoteJid:'
);

fs2.writeFileSync('/var/www/integrasi-wa.jodyaryono.id/index.js', patched);
console.log('Patch applied successfully');

File: tmp_patch_media.py
import sys

filepath = '/var/www/integrasi-wa.jodyaryono.id/index.js'

with open(filepath, 'r') as f:
    content = f.read()

old = """        const contact = await msg.getContact();
        const pushName = contact?.pushname || null;
        const payload = {
            phone_id: phoneId, message_id: msg.id?._serialized, message: textContent,
            type: contentType?.replace('Message', '') || 'text',
            timestamp: msg.timestamp || Math.floor(Date.now() / 1000),
            sender: fromNum, sender_name: pushName, from: jid.replace('@c.us', '').replace('@g.us', ''),
            pushname: pushName,
            ...(isGroup ? { group_id: jid, from_group: jid, group_name: groupCache.get(jid)?.subject || jid } : {}),
            _key: { remoteJid: jid, id: msg.id?._serialized, fromMe: msg.fromMe || false, participant: msg.author || undefined },
        };"""

new = """        const contact = await msg.getContact();
        const pushName = contact?.pushname || null;
        // Download and save media if present
        let savedMediaUrl = null;
        if (msg.hasMedia) {
            try {
                const media = await msg.downloadMedia();
                if (media?.data) {
                    const rawExt = (media.mimetype || '').split('/')[1]?.split(';')[0] || 'bin';
                    const ext = rawExt === 'jpeg' ? 'jpg' : rawExt;
                    const fname = 'wa_' + Date.now() + '_' + Math.random().toString(36).slice(2, 8) + '.' + ext;
                    const uploadDir = '/var/www/integrasi-wa.jodyaryono.id/uploads';
                    if (!fs.existsSync(uploadDir)) fs.mkdirSync(uploadDir, { recursive: true });
                    fs.writeFileSync(path.join(uploadDir, fname), Buffer.from(media.data, 'base64'));
                    savedMediaUrl = 'https://integrasi-wa.jodyaryono.id/uploads/' + fname;
                    console.log('[Media] saved ' + fname + ' (' + media.mimetype + ')');
                }
            } catch (me) { console.error('[Media] download failed', me.message); }
        }
        const payload = {
            phone_id: phoneId, message_id: msg.id?._serialized, message: textContent,
            type: contentType?.replace('Message', '') || 'text',
            timestamp: msg.timestamp || Math.floor(Date.now() / 1000),
            sender: fromNum, sender_name: pushName, from: jid.replace('@c.us', '').replace('@g.us', ''),
            pushname: pushName,
            media_url: savedMediaUrl,
            ...(isGroup ? { group_id: jid, from_group: jid, group_name: groupCache.get(jid)?.subject || jid } : {}),
            _key: { remoteJid: jid, id: msg.id?._serialized, fromMe: msg.fromMe || false, participant: msg.author || undefined },
        };"""

if old in content:
    content = content.replace(old, new)
    with open(filepath, 'w') as f:
        f.write(content)
    print("PATCHED OK")
else:
    print("ERROR: pattern not found")
    # Print surrounding context for debugging
    idx = content.find("const pushName = contact?.pushname")
    print("Context around pushName:", repr(content[idx-50:idx+200]))
    sys.exit(1)
